sql/2022/fonts/usage_of_system_families.sql

```diff
@@ -0,0 +1,311 @@
+CREATE TEMPORARY FUNCTION getSystemFamilies(data STRING)
+RETURNS ARRAY < STRING > LANGUAGE js AS '''
+// Font Family parser taken from https://github.com/bramstein/css-font-parser (BSD license)
```

Comment:
Oh this is cool! Let's add this library to [sql/lib](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/lib) and publish it on GCS so we can include it in queries more easily. Similar to how we include CSS utils in many of the CSS queries ([example](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2022/css/all_properties.sql#L5)).

That would help decouple the analysis code from the library code in this query.

Reply:
Oh, I didn't know that exists! I can add that in a separate PR. How do I deploy that library? It looks like there is a utility script to copy the files to a bucket, but I assume that requires credentials.

Reply:
One of the project maintainers would need to deploy it, but if you check the library into this PR I can stage it on GCS and show how to integrate it with the query.

Reply:
I added both libraries to this PR (one for parsing font-family/font properties, and the other for detecting writing scripts.) Let me know how I can help with staging it on GCS.

Reply:
Ok I've pushed the lib files onto GCS. IIUC there's still a bit of code needed to use the library, so the final query might look like the snippet below. All I've done is add `OPTIONS (library = "gs://httparchive/lib/css-font-parser.js")` and remove any boilerplate provided by the library.

```sql
CREATE TEMPORARY FUNCTION getSystemFamilies(data STRING)
RETURNS ARRAY < STRING > LANGUAGE js
OPTIONS (library = "gs://httparchive/lib/css-font-parser.js")
AS '''
function parseCss(rule, acc) {
  if (rule.hasOwnProperty('rules')) {
    rule.rules.forEach(rule => parseCss(rule, acc));
  }

  if (rule.hasOwnProperty('declarations')) {
    rule.declarations.forEach(decl => {
      if (decl.property === 'font-family') {
        const fonts = parseFontFamily(decl.value);

        if (fonts) {
          fonts.forEach(font => acc.push(font));
        }
      } else if (decl.property === 'font') {
        const value = parseFont(decl.value);

        if (value) {
          value['font-family'].forEach(font => acc.push(font));
        }
      }
    });
  }
}

const system = [
  'serif',
  'sans-serif',
  'cursive',
  'fantasy',
  'monospace',
  'system-ui',
  'emoji',
  'math',
  'fangsong',
  'ui-serif',
  'ui-sans-serif',
  'ui-monospace',
  'ui-rounded'
];

function getSystemFonts(json) {
  try {
    const css = JSON.parse(json);
    const result = [];

    parseCss(css.stylesheet, result);

    return result.filter(font => system.includes(font));
  } catch (e) {
    return [];
  }
}

return getSystemFonts(data);
''';

SELECT
  client,
  font_family,
  pages,
  total,
  pages / total AS pct
FROM (
  SELECT
    client,
    font_family,
    COUNT(DISTINCT page) AS pages
  FROM
    `httparchive.almanac.parsed_css`,
    UNNEST(getSystemFamilies(css)) AS font_family
  WHERE
    date = '2022-07-01'
  GROUP BY
    client,
    font_family)
JOIN (
  SELECT
    _TABLE_SUFFIX AS client,
    COUNT(0) AS total
  FROM
    `httparchive.summary_pages.2022_07_01_*`
  GROUP BY
    client)
USING
  (client)
ORDER BY
  pct DESC
```

Reply:
Thanks, I'll update the queries to use the libraries.